src/ripple/app/tx/impl/CashCheck.cpp

```diff
@@ -0,0 +1,412 @@
+                transferXRP (psb, srcId, account_, xrpDeliver, viewJ);
+            }
+        }
+        else
+        {
+            // Let flow() do the heavy lifting on a check for an IOU.
+            auto const optDeliverMin = ctx_.tx[~sfDeliverMin];
+            STAmount const flowDeliver {[&optDeliverMin] (STTx const& tx)
+            {
+                // If the check cash specified Amount deliver exactly that.
+                if (! optDeliverMin)
+                    return static_cast<STAmount>(tx[sfAmount]);
+
+                // We can't use the maximum possible currency here because
+                // there might be a gateway transfer rate to account for.
+                // Since the transfer rate cannot exceed 200%, we use 1/2
+                // maxValue for our limit.
+                return STAmount { optDeliverMin->issue(),
+                    STAmount::cMaxValue / 2, STAmount::cMaxOffset };
+            }(ctx_.tx)};
+
+            // Call the payment engine's flow() to do the actual work.
+            auto const result = flow (psb, flowDeliver, srcId, account_,
+                STPathSet{},
+                true,                              // default path
+                static_cast<bool>(optDeliverMin),  // partial payment
+                true,                              // owner pays transfer fee
+                false,                             // offer crossing
+                boost::optional<Quality>{},
+                sleCheck->getFieldAmount (sfSendMax),
+                viewJ);
+
+            if (result.result() != tesSUCCESS)
+            {
+                JLOG(ctx_.journal.warn())
+                    << "flow failed when cashing check.";
+                return result.result();
+            }
+
+            // Make sure that deliverMin was satisfied.
+            if (optDeliverMin && result.actualAmountOut < *optDeliverMin)
+            {
+                JLOG(ctx_.journal.warn()) << "flow did not produce DeliverMin.";
+                return tecPATH_PARTIAL;
+            }
+        }
+    }
+
+    // Check was cashed.  If not a self send (and it shouldn't be), remove
+    // check link from destination directory.
+    if (srcId != account_)
+    {
+        std::uint64_t const page {(*sleCheck)[sfDestinationNode]};
+        TER const ter {dirDelete (psb, true, page,
+            keylet::ownerDir (account_), checkKey, false, page == 0, viewJ)};
+        if (! isTesSuccess (ter))
+        {
+            JLOG(j_.warn()) << "Unable to delete check from destination.";
+            return ter;
+        }
+    }
+    // Remove check from check owner's directory.
+    {
+        std::uint64_t const page {(*sleCheck)[sfOwnerNode]};
+        TER const ter {dirDelete (psb, true, page,
+            keylet::ownerDir (srcId), checkKey, false, page == 0, viewJ)};
+        if (! isTesSuccess (ter))
+        {
+            JLOG(j_.warn()) << "Unable to delete check from owner.";
+            return ter;
```

Comment:
This point is not covered by the unit test. Is it worth adding?

Reply:
I don't think it's worth worrying about that.